### PR TITLE
fix(inpost): flatten nested ShipX field-errors so target_point rejections classify correctly

### LINKED ISSUE #1803: [BUG] Integration/Frontend — InPost paczkomat label generation fails with opaque ShipX validation error
> Tracking parent. Split into two native sub-issues from one report:
> - #1806 - Part A (bug): surface ShipX per-field validation details in the label-error Alert.
> - #1807 - Part B (spike): confirm and fix the underlying ShipX rejection of Allegro paczkomat labels.
>
> Full diagnosis retained below as shared context.

## Problem / Context

Generating an InPost label for a **paczkomat** shipment sourced from an Allegro order (delivery method "Allegro Paczkomaty InPost") fails. The operator sees only:

> There are some validation errors. Check details object for more info.

That string is **not** produced by OpenLinker. It is the verbatim top-level `message` of a ShipX `validation_failed` (HTTP 400) response. The actionable per-field reasons live in the ShipX `details` object, which OL captures server-side but the operator never sees. So we have two intertwined problems - a diagnostics gap that hides the cause, and the underlying ShipX rejection itself:

**1. The `details` object is captured server-side but dropped in the UI.**
The failing field(s) are available end-to-end on the backend but the FE Alert only renders the generic message:
- HTTP client attaches ShipX `details` to the exception - `libs/integrations/inpost/src/infrastructure/http/inpost-http-client.ts:192-197` (`{ fieldErrors: errorBody.details }`).
- Dispatch service logs the full structured detail - `libs/core/src/shipping/application/services/shipment-dispatch.service.ts:346-354` (`details=${JSON.stringify(error.providerDetails ?? null)}`, ref #1428), but persists only the generic `errorMessage` (lines 357-361).
- Controller returns it verbatim in the 502 body - `apps/api/src/shipping/http/shipment.controller.ts:420-429` (`details: error.providerDetails`).
- **FE loses it:** `ApiError.fromResponse` stores the whole body on `.details` (`apps/web/src/shared/api/api-error.ts:36-46`), but `GenerateLabelForm` renders only `mutation.error.message` in its Alert (`apps/web/src/features/orders/components/generate-label-form.tsx:477-481`). The `details.fieldErrors` payload is present on `mutation.error.details` and never displayed.

**2. Likely underlying ShipX rejection: receiver fields on the locker path.**
For a paczkomat, the request builder omits the receiver address (correct) but forwards `company_name = recipient.name`, `first_name`, `last_name`, `email`, `phone` - `libs/integrations/inpost/src/infrastructure/mappers/inpost-shipx.mapper.ts:446-459`. The FE builder never populates `recipient.name`, and the FE preflight gate `detectMissingFields` only requires `customerEmail` + `phone` for lockers (`apps/web/src/features/orders/lib/dispatch-input.ts:65-87`); it never checks `firstName`/`lastName`. Allegro buyer snapshots frequently lack a parsed first/last name, so an order can pass OL's FE gate and then be rejected by ShipX on the `receiver`. This is the strongest candidate but **must be confirmed** against a live `details` payload (see Spike below), because ShipX could equally be rejecting `parcels.template`, `custom_attributes.target_point`, or a sender field.

## Proposed Solution

Two deliverables - a quick FE diagnostics fix, then a spike gated on what it reveals.

**Part A - Bug fix (surface ShipX `details` in the label-error Alert).** In `GenerateLabelForm`, when the mutation error carries a structured `details.fieldErrors` (the ShipX `details` map of `field -> string[]`), render the per-field messages under the generic Alert message instead of only `mutation.error.message`. Read the shape off `ApiError.details` (`apps/web/src/shared/api/api-error.ts`). This immediately turns the opaque failure into an actionable one and is the diagnostic that confirms Part B's root cause. Keep it generic (render whatever `fieldErrors` map ShipX returns), not paczkomat-specific.

**Part B - Spike + targeted fix (confirm and close the real rejection).** Using the now-visible `details` (or the existing backend WARN log at `shipment-dispatch.service.ts:352`), confirm which ShipX field is rejected for an Allegro paczkomat order. If it is the receiver (most likely), fix the FE preflight (`detectMissingFields`) and/or the mapper (`toReceiverPeer`) so a locker shipment always sends an identifiable receiver ShipX accepts (e.g. fall back `company_name`/`first_name`/`last_name` from the Allegro snapshot buyer name, or gate submission with a clear operator-facing message when the name is genuinely absent). Follow the existing preflight-guard pattern in `inpost-shipx.mapper.ts:298-312`. Add the field-level requirement to the InPost package docs.

## Classification

**Type**: Integration + Frontend
**Layer**: Infrastructure (InPost adapter/mapper) + Interface (FE label form) + Application (FE preflight)
**File(s)**:
- `apps/web/src/features/orders/components/generate-label-form.tsx` (Alert - Part A)
- `apps/web/src/shared/api/api-error.ts` (details shape)
- `apps/web/src/features/orders/lib/dispatch-input.ts` (FE preflight - Part B)
- `libs/integrations/inpost/src/infrastructure/mappers/inpost-shipx.mapper.ts` (receiver mapping - Part B)
- `libs/integrations/inpost/src/infrastructure/http/inpost-http-client.ts` (details attach - reference)

## Docs impact

- **Package-local**: `libs/integrations/inpost/README.md` / `docs/setup-guide.md` - if Part B lands a receiver-field requirement for locker shipments, document which receiver fields ShipX requires for a paczkomat (currently undocumented; contract notes live only in code comments at `inpost-shipx.mapper.ts:319-323`).
- **`docs/lessons.md`**: if the root cause is the FE-preflight vs ShipX-required-fields mismatch, this is a recurring empirical pitfall worth recording (OL preflight passing a request that the provider then rejects).
- **Central reference docs**: None expected - no new port, capability, or cross-context edge.
- **ADR**: None expected - no architectural decision with cross-context trade-offs.

## Dependencies

- None. Part A is independent; Part B is gated on Part A's output (or the existing backend log) but needs no other issue merged first.

## Assumptions

- The reported failure is a ShipX `validation_failed` (400), consistent with the exact generic message shown. The specific rejected field is unconfirmed until the `details` payload is inspected - hence the spike half.
- The receiver-name gap is the leading hypothesis but not assumed proven; Part B explicitly confirms before fixing.
- This is not a credential/auth failure (that surfaces differently via the auth-failure classifier, not `validation_failed`).

## Acceptance Criteria

- [ ] `GenerateLabelForm` renders the ShipX per-field `details`/`fieldErrors` messages in the error Alert when present, falling back to the generic message when absent.
- [ ] The exact ShipX field(s) rejected for an Allegro paczkomat order are identified and recorded in this issue (or its PR).
- [ ] The confirmed root cause is fixed (receiver-field mapping/preflight, or whichever field the spike identifies) so an Allegro paczkomat label generates successfully end to end.
- [ ] FE preflight and the ShipX request builder no longer allow a locker request that ShipX will reject for a missing required field, OR the operator gets a clear pre-submit message naming the missing data.
- [ ] Tests added or updated for non-trivial logic (FE Alert rendering of field errors; mapper/preflight receiver handling).
- [ ] No architecture boundary violations (CORE <-> Integration).
- [ ] Documentation updated per **Docs impact** (InPost package docs + lessons.md if applicable), or explicitly confirmed none.

### LINKED ISSUE #1807: [SPIKE] Integration/Frontend — Confirm and fix ShipX rejection of Allegro paczkomat labels
> Sub-issue of #1803. This is Part B (the spike + targeted fix). It confirms and closes the underlying ShipX rejection for Allegro paczkomat labels. Gated on the field-level errors surfaced by the Part A diagnostics bug (or the existing backend WARN log).

## Problem / Context

Generating an InPost label for a **paczkomat** shipment sourced from an Allegro order ("Allegro Paczkomaty InPost") is rejected by ShipX with `validation_failed`. The exact field is not yet confirmed - the leading hypothesis is the **receiver** on the locker path:

- For a paczkomat the mapper omits the receiver address (correct) but forwards `company_name = recipient.name`, `first_name`, `last_name`, `email`, `phone` - `libs/integrations/inpost/src/infrastructure/mappers/inpost-shipx.mapper.ts:446-459`.
- The FE builder never populates `recipient.name`, and the FE preflight `detectMissingFields` requires only `customerEmail` + `phone` for lockers - `apps/web/src/features/orders/lib/dispatch-input.ts:65-87`; it never checks `firstName`/`lastName`.
- Allegro buyer snapshots frequently lack a parsed first/last name, so an order passes OL's FE gate and is then rejected by ShipX on `receiver`.

This is the strongest candidate but is **not proven** - ShipX could equally be rejecting `parcels.template`, `custom_attributes.target_point`, or a sender field. The spike confirms before fixing.

## Proposed Solution

**Spike:** Using the now-visible ShipX `details` (Part A) or the backend WARN log at `libs/core/src/shipping/application/services/shipment-dispatch.service.ts:352`, confirm which ShipX field is rejected for an Allegro paczkomat order. Record the finding in this issue.

**Targeted fix (shape depends on the spike):** If it is the receiver (most likely), fix the FE preflight (`detectMissingFields`) and/or the mapper (`toReceiverPeer`) so a locker shipment always sends an identifiable receiver that ShipX accepts - e.g. derive `company_name`/`first_name`/`last_name` from the Allegro snapshot buyer name, or gate submission with a clear operator-facing message when the name is genuinely absent. Follow the existing preflight-guard pattern at `inpost-shipx.mapper.ts:298-312`.

## Classification

**Type**: Integration + Frontend
**Layer**: Infrastructure (InPost mapper) + Application (FE preflight)
**File(s)**:
- `libs/integrations/inpost/src/infrastructure/mappers/inpost-shipx.mapper.ts` (receiver mapping, preflight guards)
- `apps/web/src/features/orders/lib/dispatch-input.ts` (FE preflight)
- `libs/integrations/inpost/src/infrastructure/http/inpost-http-client.ts` (details attach - reference)

## Docs impact

- **Package-local**: `libs/integrations/inpost/README.md` / `docs/setup-guide.md` - document which receiver fields ShipX requires for a paczkomat if the fix lands a new requirement (currently undocumented; contract notes live only in code comments at `inpost-shipx.mapper.ts:319-323`).
- **`docs/lessons.md`**: if the root cause is the FE-preflight vs ShipX-required-fields mismatch, record it as a recurring empirical pitfall (OL preflight passing a request the provider then rejects).
- **Central reference docs / ADR**: None expected.

## Dependencies

- Gated on #1806 (the surfaced `details` make the field identifiable). The existing backend WARN log is a fallback, so this can start before #1806 merges if needed.

## Assumptions

- The failure is a ShipX `validation_failed` (400), not a credential/auth failure.
- The receiver-name gap is the leading hypothesis, not assumed proven - the spike confirms before any fix.

## Acceptance Criteria

- [x] The exact ShipX field(s) rejected for an Allegro paczkomat order are identified and recorded in this issue (`custom_attributes.target_point`, code `does_not_exist` — see comments).
- [x] The confirmed root cause (a shape-mismatch bug that broke the #885 `target_point` classification) is fixed. Note: a label for a *specific* nonexistent/decommissioned pickup-point id will still be rejected by ShipX — that's provider-side data reality, not something OL code can force through; the fix ensures that rejection is now correctly classified instead of generic.
- [x] Not a missing-field scenario (disproven receiver-name hypothesis), so no FE preflight change was needed; the ShipX request builder / error classification now correctly names the actual rejected field (`target_point`) for downstream display.
- [x] Tests added for the actual fix location (`InpostHttpClient` field-error flattening) reproducing the live nested ShipX shape.
- [x] No architecture boundary violations (CORE <-> Integration) — change is confined to the InPost integration package.
- [x] Documentation updated: `libs/integrations/inpost/docs/setup-guide.md` + `InpostConnectionTesterAdapter` docblock. No `docs/lessons.md` entry — see PR #1816 doc-impact note for why.

### PR DESCRIPTION
**Pull request by norbert-kulus-blockydevs**
## Summary

Spike (#1807) confirming and fixing the ShipX `validation_failed` rejection for Allegro-sourced InPost paczkomat labels. **The leading hypothesis (missing receiver name) was disproven** by live reproduction; this fixes the confirmed real root cause instead.

## Confirmed root cause

Reproduced live against the shared dev stack by re-issuing `generate-label` for a real failed shipment. ShipX's actual response:

```json
{
  "message": "There are some validation errors. Check details object for more info.",
  "providerCode": "custom_attributes",
  "details": { "fieldErrors": { "custom_attributes": [ { "target_point": ["does_not_exist"] } ] } }
}
```

- The rejected field is `custom_attributes.target_point` (the paczkomat/pickup-point id), **not the receiver** — the reproduced order snapshot did carry `firstName`/`lastName`, and those correctly flow into ShipX's `receiver` object already.
- `InpostShippingAdapter` already has logic from #885 that re-tags a `target_point` rejection with a stable `providerCode: 'target_point'` (so callers can offer "pick another locker"). It never fired here because ShipX returns this specific field error **nested** (`{ custom_attributes: [{ target_point: [...] }] }`), not the **flat** `{ field: [...] }` shape `InpostHttpClient`/`ShipXErrorBody.details` assumed. The outer key (`custom_attributes`) got picked as `providerCode` instead of the real offending field, and the adapter's `target_point`-key lookup missed it too — defeating the #885 classification for exactly the case it was built for.

## Fix

- `libs/integrations/inpost/src/infrastructure/http/inpost-http-client.ts` — `flattenShipXFieldErrors` normalises both the flat and nested-array ShipX error shapes onto the leaf field key before `firstDetailKey`/`providerDetails.fieldErrors` are built, so `target_point` rejections classify correctly regardless of which shape ShipX used for that field.
- `libs/integrations/inpost/src/domain/types/inpost-shipx.types.ts` — widened `ShipXErrorBody.details` to the real, confirmed union shape.
- `libs/integrations/inpost/src/infrastructure/http/__tests__/inpost-http-client.spec.ts` — new test reproducing the exact live nested shape (`custom_attributes: [{ target_point: [...] }]`) and asserting it now classifies as `providerCode: 'target_point'`.

### Health-check follow-up (requested during the spike)

Investigated whether the InPost connection-tester ("Test Connection") could produce a false positive while label generation genuinely fails. Confirmed: `InpostConnectionTesterAdapter` only calls `GET /v1/points?per_page=1` (auth/connectivity only) and never exercises shipment creation, so it structurally cannot catch a `target_point`-class rejection — live-verified the connection reports `success: true` while the paczkomat shipment above is failing. Not changing the health check to create real throwaway shipments (ShipX has no side-effect-free "validate this shipment" endpoint, and doing so would create real, in production billable, carrier-side artifacts on every connection test) — documented the scope/limit explicitly instead, in the adapter docblock and the setup guide.

Also checked whether this looks like general sandbox flakiness (per user report of "worked, then stopped, no code change"): the failure/success pattern tracks the **specific paczkomat id**, not calendar time (some ids succeed, others fail with `does_not_exist`, interleaved over the same days — including one id rejected identically on two different days). Documented as the expected explanation, distinct from a regression.

## Docs

- `libs/integrations/inpost/docs/setup-guide.md` — new Troubleshooting entries: `target_point does_not_exist` (with the confirmed nested-shape detail), the receiver-field reality (email+phone required, name not), the connection-test scope limitation, and the per-point (not time-based) "flakiness" explanation.
- `libs/integrations/inpost/src/infrastructure/adapters/inpost-connection-tester.adapter.ts` — docblock states the health-check's scope explicitly.
- No `docs/lessons.md` entry — the actual cause (shape-parsing bug at a specific integration boundary) doesn't generalize into a cross-cutting regression-ledger pattern the way the receiver-name hypothesis would have.
- No FE change: the confirmed root cause is not a preflight-checkable missing field, so `dispatch-input.ts` needed no edit (`generate-label-form.tsx` is out of scope for this PR, owned by #1806).

## Test plan

- [x] `pnpm --filter @openlinker/integrations-inpost test` — 137/137 passing (new test included)
- [x] `pnpm --filter @openlinker/integrations-inpost lint` — clean
- [x] `pnpm --filter @openlinker/integrations-inpost type-check` — clean
- [x] Live-reproduced the exact ShipX rejection against the shared dev stack and confirmed the fixed classification logic against the real response shape

Closes #1807
Closes #1803

### CODE REVIEW

File: libs/integrations/inpost/src/infrastructure/http/inpost-http-client.ts [hunk 1]
```diff
@@ -260,3 +262,48 @@ function firstDetailKey(
...
+      continue;
+    }
+    for (const item of value) {
+      if (item === null || typeof item !== 'object') {
+        continue;
+      }
+      const nestedFieldErrors = item as Record<string, unknown>;
+      for (const [nestedKey, nestedValue] of Object.entries(nestedFieldErrors)) {
+        if (Array.isArray(nestedValue) && nestedValue.every((m) => typeof m === 'string')) {
+          flat[nestedKey] = nestedValue as readonly string[];
```
Comment (piotrswierzy) [inline libs/integrations/inpost/src/infrastructure/http/inpost-http-client.ts]: SUGGESTION (non-blocking) — LINE: leaf-key flattening is last-write-wins, so it can silently drop messages on a genuine key collision. Two cases: (1) the same leaf key repeated across array items, e.g. `parcels: [{ weight: ["required"] }, { weight: ["too_heavy"] }]` — only the second `weight` survives; (2) a flat top-level key and a nested leaf sharing a name, e.g. `{ target_point: ["a"], custom_attributes: [{ target_point: ["b"] }] }` — order-dependent overwrite. For this PR's purpose it's harmless: `firstDetailKey` only needs *a* code, and the confirmed `custom_attributes` payload is a single-element array, so `target_point` is unique. Worth either a one-line comment noting the intentional flatten-collapse, or merging colliding messages (`flat[k] = [...(flat[k] ?? []), ...v]`) if you want the FE's `StructuredErrorList` (#1812) to show every rejected sub-field rather than the last one. Not blocking.

File: libs/integrations/inpost/src/infrastructure/http/inpost-http-client.ts [hunk 2]
```diff
@@ -260,3 +262,48 @@ function firstDetailKey(
...
+    if (value.every((item): item is string => typeof item === 'string')) {
+      flat[key] = value;
+      continue;
+    }
+    for (const item of value) {
+      if (item === null || typeof item !== 'object') {
+        continue;
+      }
+      const nestedFieldErrors = item as Record<string, unknown>;
+      for (const [nestedKey, nestedValue] of Object.entries(nestedFieldErrors)) {
```
Comment (piotrswierzy) [inline libs/integrations/inpost/src/infrastructure/http/inpost-http-client.ts:315]: NOTE (non-blocking): the flatten handles exactly one level of array-of-objects nesting. A doubly-nested `details` (`{ a: [{ b: [{ c: [...] }] }] }`) would be silently dropped — the inner `nestedValue` is an array of objects, fails the `every(string)` guard, and nothing is emitted. Given ShipX's two confirmed live shapes only go one level deep (#1807), this is acceptable and I'd rather not over-generalize to arbitrary recursion for shapes that don't exist. Flagging only so it's a conscious boundary. Defensive parsing here is otherwise solid — `!details` guard, non-array skip, null/non-object item guard, `unknown` + narrowing (no `any`), no throw on malformed input.

File: libs/integrations/inpost/src/infrastructure/http/__tests__/inpost-http-client.spec.ts [hunk 3]
```diff
@@ -144,6 +144,32 @@ describe('InpostHttpClient', () => {
     expect(fetchMock).toHaveBeenCalledTimes(1);
   });
 
+  it('should flatten a nested ShipX field-error map (custom_attributes.target_point) onto its leaf key, not the outer field (#1807)', async () => {
```
Comment (piotrswierzy) [inline libs/integrations/inpost/src/infrastructure/http/__tests__/inpost-http-client.spec.ts:147]: SUGGESTION (non-blocking): the new test nails the exact live nested shape, which is the important case. Since the value of this fix is largely defensive normalisation, consider one more cheap unit test locking the malformed/edge paths the function is written to survive — e.g. a nested item that is `null` or a non-array leaf value (`{ custom_attributes: [{ target_point: "oops" }] }`) — asserting it doesn't throw and simply omits the unparseable key. Locks the "no throw on garbage payload" contract so a future refactor can't regress it. Not blocking.

**Review by piotrswierzy — APPROVED** [review]
## Tech Lead review — APPROVE ✅

Clean, well-scoped fix with an excellent PR writeup (disproven hypothesis + live-confirmed root cause + per-point-id vs time-based flakiness analysis). Verified against both contract endpoints; approving with suggestions only (none blocking).

### Correctness — the flatten is right
`flattenShipXFieldErrors` correctly normalises both confirmed ShipX shapes onto the **leaf key**:
- flat `{ name: ["required"] }` → `{ name: [...] }`
- nested `{ custom_attributes: [{ target_point: ["does_not_exist"] }] }` → `{ target_point: [...] }`

Crucially it emits the **leaf key, not a dotted path** — which is exactly what the existing #885 re-tag needs: `InpostShippingAdapter.mentionsTargetPoint` does `Object.keys(fieldErrors).includes('target_point')`. A dotted `custom_attributes.target_point` would have *broken* that lookup. So the design choice here is correct, not a shortcut.

### Neutral-shape agreement with FE #1812 — confirmed
FE `extractShippingFieldErrors` consumes `ApiError.details.fieldErrors` as a flat `Record<field, string[]>` and renders one `StructuredErrorList` row per field+message, tagged by `providerCode`. The BE now guarantees that flat leaf-keyed shape. Pre-fix, the nested array-of-objects would have reached the FE as non-`string[]` values and broken rendering. The two PRs agree on the wire contract.

### Boundary hygiene
Mapping lives entirely in the InPost plugin's infrastructure/http layer → neutral `ShippingProviderRejectionException` (`providerCode` + `providerDetails.fieldErrors`); no ShipX vocabulary leaks into core. `ShipXErrorBody.details` widened to the real union. No `any` — `unknown` + narrowing on the external payload. No secrets. Defensive parsing is solid (null/non-array/non-object all guarded; no throw on malformed input). Test reproduces the exact live shape.

### Non-blocking suggestions (inline)
1. Leaf-key flatten is last-write-wins → silent message loss on a genuine key collision (multi-item arrays / flat+nested name clash). Harmless for this PR's single-element `custom_attributes` + `firstDetailKey` usage; consider a comment or message-merge if you want the FE to show every colliding sub-field.
2. One-level-deep nesting only; deeper shapes silently dropped — acceptable given the two confirmed live shapes, flagged as a conscious boundary.
3. Consider one more unit test locking the malformed/edge paths (null nested item, non-array leaf) to protect the "no throw on garbage" contract.

Non-code call-outs are good judgment: correctly declined to make "Test Connection" create billable throwaway shipments, and correctly declined a `docs/lessons.md` entry (shape-parse bug at one boundary doesn't generalize). Ship it.

**Comment by norbert-kulus-blockydevs** [comment]
Thanks for the review. Addressed all three suggestions in 94924ae9.

**1. Last-write-wins on colliding leaf keys** - went with the message-merge option rather than just a comment, so the FE `StructuredErrorList` (#1812) shows every rejected sub-field:

```ts
const mergeInto = (fieldKey: string, messages: readonly string[]): void => {
  flat[fieldKey] = [...(flat[fieldKey] ?? []), ...messages];
};
```

Both the flat and nested branches now route through `mergeInto`, so `parcels: [{ weight: ["required"] }, { weight: ["too_heavy"] }]` yields `weight: ["required", "too_heavy"]` and a flat+nested name clash accumulates instead of overwriting.

**2. One-level-deep nesting boundary** - kept the boundary as-is (agree on not over-generalising to arbitrary recursion for shapes ShipX doesn't emit), but made it an explicit, conscious non-goal in the function docblock rather than silent behaviour: a doubly-nested `details` fails the string-array guard and is dropped, not recursed.

**3. Malformed/edge-path test** - added two unit tests locking the defensive contract:
- message merge across repeated leaf keys (case 1 above);
- a malformed payload combining a `null` nested item, a non-array leaf (`{ target_point: "oops" }`), and a doubly-nested key - asserting it doesn't throw and simply omits the unparseable keys while the one well-formed leaf survives.

`pnpm --filter @openlinker/integrations-inpost test` - 139/139 passing; lint + type-check clean.
